lizrd/core/distributed.py

```diff
@@ -0,0 +1,94 @@
+from typing import Optional
+
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.distributed.fsdp import MixedPrecision, CPUOffload
+from torch.nn.parallel import DistributedDataParallel as DDP
+import torch.nn as nn
+import torch
+
+from lizrd.core.misc import Noop
+
+
+def wrap_module_in_fsdp(
+    enabled: bool,
+    module: nn.Module,
+    rank: Optional[int],
+    param_precision: torch.dtype,
+    cast_inputs: bool = False,
+    offload_params: bool = False,
+    print_model: bool = False,
+):
+    wrapped = FSDP(
+        module,
+        device_id=rank,
+        mixed_precision=MixedPrecision(
+            param_dtype=param_precision,
+            reduce_dtype=torch.float32,
+            cast_forward_inputs=cast_inputs,
+        ),
+        cpu_offload=CPUOffload(offload_params=offload_params),
+    )
+    if print_model:
+        print("------- MODEL AFTER WRAPPING IN FSDP -------")
+        print(wrapped)
+        print("--------------------------------------------")
+    return wrapped
+
+
+def wrap_in_fsdp(
+    enabled: bool,
+    module: nn.Module,
+    rank: Optional[int],
+    param_precision: torch.dtype,
+    cast_inputs: bool = False,
+    offload_params: bool = False,
+    print_model: bool = False,
+    cast_outputs_to: Optional[torch.dtype] = None,
+):
+    if not enabled:
+        return module
```

Comment:
I think we should not wrap all possible code with a wrapper that sometimes does nothing. The code clarity will benefit if we use the wrap when we actually intend to wrap something, and not use it otherwise.
This, of course, comes at an expense of more _if_ statements, but I think it makes the code less abstract and easier to understand.

Reply:
I see a future where we have a wrapper called sth like `wrap_if_needed` which is basically: 
`if no wrap: return module, elif ddp: return ddp_wrap(module), elif fsdp: return fsdp` and so on

Reply:
I think this wrapper is really straightforward and accepts `enabled` as the first argument. The code is much more readable without a lot of identical `if` statements

Reply:
I really would like the `wrap_if_needed` implemented. Hiding the context in the arguments is not clear enough, in my opinion. Also, a person (e.g. me) that does not know what FSDP does exactly might not know that `enabled` means that **all functionality** is disabled.

Reply:
Let's meet offline and discuss this matter. I can potentially change the name of this function to `wrap_if_needed` too